posts/2024-11-28-Rust-1.83.0.md

```diff
@@ -0,0 +1,96 @@
+---
+layout: post
+title: "Announcing Rust 1.83.0"
+author: The Rust Release Team
+release: true
+---
+
+The Rust team is happy to announce a new version of Rust, 1.83.0. Rust is a programming language empowering everyone to build reliable and efficient software.
+
+If you have a previous version of Rust installed via `rustup`, you can get 1.83.0 with:
+
+```console
+$ rustup update stable
+```
+
+If you don't have it already, you can [get `rustup`](https://www.rust-lang.org/install.html) from the appropriate page on our website, and check out the [detailed release notes for 1.83.0](TODO).
+
+If you'd like to help us out by testing future releases, you might consider updating locally to use the beta channel (`rustup default beta`) or the nightly channel (`rustup default nightly`). Please [report](https://github.com/rust-lang/rust/issues/new/choose) any bugs you might come across!
+
+## What's in 1.83.0 stable
+
+### New const capabilities
+
+This release includes several large extensions to what code running in const contexts can do. This refers to all code that the compiler has to evaluate at compile-time: the initial value of `const` and `static` items, array lengths, enum discriminant values, const generic arguments, and functions callable from such contexts (`const fn`).
+
+**References to statics.**
+So far, `const` items and `const fn` were forbidden from referencing `static items`.
+This limitation has now been lifted:
+```rust
+static S: i32 = 25;
+const C: &i32 = &S;
+```
+Note, however, that reading the value of a mutable or interior mutable static is still not permitted in const contexts. Furthermore, the final value of a constant may not reference any mutable or interior mutable statics:
```

Comment:
```suggestion
Note, however, that reading the value of a mutable or interior mutable static is still not permitted in const contexts (except static initializers). Furthermore, the final value of a constant may not reference any mutable or interior mutable statics:
```

This is a nit, and I could understand not even saying it.  It's related to what we're documenting here:

https://github.com/rust-lang/reference/pull/1657/files

cc @RalfJung

Reply:
Maybe we should say something like "still not permitted in const initializers"?

Reply:
As far as I know, static initializers are the only exception (that is, they are a [const contexts](https://doc.rust-lang.org/nightly/reference/const_eval.html?highlight=const#const-context) but do allow this), and reading such a value in all other const contexts is not permitted.

As one maybe surprising example of this, we accept...

```rust
static S: u8 = {
    static mut S: &mut u8 = &mut 42;
    unsafe { *S }
};
```

...but reject:

```rust
static S: u8 = {
    static mut S: &mut u8 = &mut 42;
    unsafe { const { *S } }
    //~^ ERROR evaluation of `S::{constant#0}` failed
};
```

Reply:
Yeah, but the same is true for the statement above the example which says that so far, mentioning statics is not allowed. There we say "`const` items and `const fn` were forbidden from referencing" but that's really just a different way of saying "const contexts except for statics".

Reply:
There's a separate distinction about mutable statics too though, no?  E.g., we accept this...

```rust
static S: u8 = {
    static S: u8 = 42;
    const { S }
};
```

...but not this:

```rust
static S: u8 = {
    static mut S: u8 = 42;
    unsafe { const { S } }
    //~^ ERROR evaluation of `S::{constant#0}` failed
};
```

Reply:
There's a separate distinction about mutable statics too though, no?  E.g., we accept this...

```rust
static S: u8 = {
    static S: u8 = 42;
    const { S }
};
```

...and this...

```rust
static S: u8 = {
    static mut S: u8 = 42;
    unsafe { S }
};
```

...but not this:

```rust
static S: u8 = {
    static mut S: u8 = 42;
    unsafe { const { S } }
    //~^ ERROR evaluation of `S::{constant#0}` failed
};
```

Reply:
(I keep getting duplicate notifications about your posts, and the first ones that get sent seem to be invalid links... something odd is going on here.)

`const { S }` is like an anonymous const item for the purpose of this, the surrounding `static` is irrelevant at this point. So `const { S }` works iff `S` is an immutable static.

Reply:
And so if we are to say...

> Note, however, that reading the value of a mutable or interior mutable static is still not permitted in const contexts.

...then it's still not exactly correct unless we say something like:

> Note, however, that reading the value of a mutable or interior mutable static is still not permitted in const contexts (except static initializers).

Reply:
> The text above, as proposed...

That describes the status quo, on stable. It says "so far" in release notes, which is intended to refer to the previous version.

At this point I am quite confused by your comments.^^

Reply:
So my proposal would be the diff I suggested to the text above this example, plus
```suggestion
Note, however, that reading the value of a mutable or interior mutable static is still not permitted in these const contexts. Furthermore, the final value of a constant may not reference any mutable or interior mutable statics:
```

One could go into a lot more detail, but these are just release notes after all.

Reply:
Talked with @RalfJung separately.  I understand now what he's going for here is the idea that by qualifying this earlier that when we say "const contexts" later that we mean it as earlier qualified.  That's probably not what I would have taken away with it as worded, but I see now the intention.  RalfJ has proposed saying something along the lines of "these const contexts", which I think would probably make this clear enough.

Reply:
I've turned this into github suggestions: https://github.com/rust-lang/blog.rust-lang.org/pull/1432#discussion_r1851456339 and https://github.com/rust-lang/blog.rust-lang.org/pull/1432#discussion_r1852713791

Reply:
> > https://github.com/rust-lang/reference/pull/1657/files
> 
> Btw I think that's the wrong PR, do you mean [rust-lang/reference#1610](https://github.com/rust-lang/reference/pull/1610) ?

I mean particularly this language being added in rust-lang/reference#1657:

> Static initializers may refer to and read from other statics.  When reading from mutable statics, they read the initial value of that static.

You're right that there is overlap between those two, as I had noted here: https://github.com/rust-lang/reference/pull/1610#discussion_r1830186152.